repos/system_upgrade/common/actors/scangrubconfig/libraries/scanner.py

```diff
@@ -0,0 +1,74 @@
+import os
+import re
+
+from leapp.libraries.common.config import architecture, version
+from leapp.models import GrubConfigError
+
+
+def is_grubenv_corrupted(conf_file):
+    # grubenv can be missing
+    if not os.path.exists(conf_file):
+        return False
+    # ignore when /boot/grub2/grubenv is a symlink to its EFI counterpart
+    if os.path.islink(conf_file) and os.readlink(conf_file) == '../efi/EFI/redhat/grubenv':
+        return False
+    with open(conf_file, 'r') as config:
+        config_contents = config.read()
+    return len(config_contents) != 1024 or config_contents[-1] == '\n'
+
+
+def _get_config_contents(config_path):
+    if os.path.isfile(config_path):
+        with open(config_path, 'r') as config:
+            return config.read()
+    return ''
+
+
+def is_grub_config_missing_final_newline(conf_file):
+    config_contents = _get_config_contents(conf_file)
+    return config_contents != '' and config_contents[-1] != '\n'
+
+
+def detect_config_error(conf_file):
+    """
+    Check grub configuration for syntax error in GRUB_CMDLINE_LINUX value.
+
+    :return: Function returns True if error was detected, otherwise False.
+    """
+    with open(conf_file, 'r') as f:
+        config = f.read()
+
+    pattern = r'GRUB_CMDLINE_LINUX="[^"]+"(?!(\s*$)|(\s+(GRUB|#)))'
+    return re.search(pattern, config) is not None
+
+
+def scan():
+    errors = []
+    # Check for corrupted grubenv
+    if not architecture.matches_architecture(architecture.ARCH_S390X):
+        configs = ['/boot/grub2/grubenv', '/boot/efi/EFI/redhat/grubenv']
+        corrupted = []
+        for cfg in configs:
+            if is_grubenv_corrupted(cfg):
+                corrupted.append(cfg)
+        if corrupted:
+            errors.append(GrubConfigError(error_detected=True, error_type=GrubConfigError.ERROR_CORRUPTED_GRUBENV,
+                                          files=corrupted))
+
+    # Check for GRUB_CMDLINE_LINUX syntax errors
+    # XXX FIXME(ivasilev) Can we make this check a common one? For now let's limit it to rhel7->rhel8 only
+    if version.get_major_version() == '7':
+        if not architecture.matches_architecture(architecture.ARCH_S390X):
+            # For now, skip just s390x, that's only one that is failing now
+            # because ZIPL is used there
+            config = '/etc/default/grub'
+            if detect_config_error(config):
+                errors.append(GrubConfigError(error_detected=True,
+                                              error_type=GrubConfigError.ERROR_GRUB_CMDLINE_LINUX_SYNTAX))
```

Comment:
(just some related thoughts on the matter)
I am very intrigued why do we keep the error_detected field in this Model. TBH I can't imagine the use for GrubConfigError with error_detected=False.
Do you have any idea?

Reply:
hmm, e.g.
```
if not error.error_detected:
    raise StopActorExecution
```
also it's easy check that we haven't detected any error inside the grub configuration when investigating. yes, we argue that the same could be achieved when the msg is not created at all, but it already exists, so are going to keep it in this way. It's not a new field so I would keep it.

Reply:
I don't see any occurencies like you mentioned in the code base unfortunately :(
The reason I am asking is because I'd like to add a docstring explaining why this field is here and what value should be set.

However I did find this in [addupgradebootentry actor](https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/common/actors/addupgradebootentry/actor.py#L23):
```
if grub_config_error.error_detected:
                fix_grub_config_error('/etc/default/grub', grub_config_error.error_type)
```
fix_grub_config implementation [is here](https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py#L99)

How about I put the following docstring
```
If the error can be automatically corrected (ex. in add upgrade boot entry actor), please set the error_detected field to True
```
?

Reply:
Hmmm now thinking about it - if that is indeed the case, how about renaming `error_detected` to `error_resolvable`?

Reply:
we cannot just rename it, the same problem why we cannot just remove it. we have the deprecation process so in case we would like to do something like that, we need to introduce the new field and support the old one for the protected period. but it's unfortunate use indeed when seeing the code.